File: server/core/src/main/java/dev/slimevr/protocol/ProtocolHandler.java
package dev.slimevr.protocol;

import java.util.function.BiConsumer;


public abstract class ProtocolHandler<H> {

	public final BiConsumer<GenericConnection, H>[] handlers;

	public ProtocolHandler() {
		this.handlers = new BiConsumer[this.messagesCount()];
	}

	public abstract void onMessage(GenericConnection conn, H message);

	public abstract int messagesCount();

	public void registerPacketListener(byte packetType, BiConsumer<GenericConnection, H> consumer) {
		if (this.handlers[packetType] != null)
			this.handlers[packetType] = this.handlers[packetType].andThen(consumer);
		else
			this.handlers[packetType] = consumer;
	}


}

File: server/core/src/main/java/dev/slimevr/protocol/ProtocolHandler.kt
package dev.slimevr.protocol

abstract class ProtocolHandler<H> {
	@JvmField
	val handlers: Array<((GenericConnection, H) -> Unit)?>

	init {
		this.handlers = arrayOfNulls(this.messagesCount())
	}

	abstract fun onMessage(conn: GenericConnection, message: H)

	abstract fun messagesCount(): Int

	fun registerPacketListener(packetType: Byte, consumer: (GenericConnection, H) -> Unit) {
		val packetInt = packetType.toInt()
		if (handlers[packetInt] != null) {
			val previous = handlers[packetInt]!!
			handlers[packetInt] = { conn, msg ->
				previous(conn, msg)
				consumer(conn, msg)
			}
		} else {
			handlers[packetInt] = consumer
		}
	}
}
